test/TypeCoercion/immutable_mutable.swift

```diff
@@ -0,0 +1,55 @@
+// RUN: %target-typecheck-verify-swift -parse-stdlib
```

Comment:
This is only testing type-checking, but based on the comment I made in CSApply, I think later compiler stages (e.g., SILGen) will fail due to an AST that doesn't have well-formed types. I recommend adding some SILGen tests to ensure that the right initializers are getting called when this conversion applies.